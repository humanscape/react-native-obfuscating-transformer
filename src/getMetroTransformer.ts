import { Node } from "babel-core"
import { RawSourceMap, SourceMapConsumer } from "source-map"
import * as semver from "semver"
import {
  MetroRawSourceMap,
  convertStandardSourceMapToMetroRawSourceMap,
} from "./composeSourceMaps"
import * as babylon from "babylon"
import traverse from "babel-traverse"

export interface MetroTransformerResult {
  ast?: Node
  code?: string
  map?: string | RawSourceMap | MetroRawSourceMap
}

export interface MetroTransformer {
  transform(props: {
    filename: string
    src: string
    options: {
      dev?: boolean
      retainLines?: boolean
      // others unused
    }
  }): MetroTransformerResult
  getCacheKey?(): string
}

function getReactNativeMinorVersion(): number {
  const reactNativeVersionString = require("react-native/package.json").version

  const parseResult = semver.parse(reactNativeVersionString)

  if (!parseResult) {
    throw new Error(
      `Can't parse react-native version string '${reactNativeVersionString}'`,
    )
  }

  return parseResult.minor
}

export function getMetroTransformer(
  reactNativeMinorVersion: number = getReactNativeMinorVersion(),
): MetroTransformer {
  if (reactNativeMinorVersion >= 59) {
    return require('metro-react-native-babel-transformer/src/index')
  } else if (reactNativeMinorVersion >= 56) {
    return require("metro/src/reactNativeTransformer")
  } else if (reactNativeMinorVersion >= 52) {
    return require("metro/src/transformer")
  } else if (reactNativeMinorVersion >= 0.47) {
    return require("metro-bundler/src/transformer")
  } else if (reactNativeMinorVersion === 0.46) {
    return require("metro-bundler/build/transformer")
  } else {
    throw new Error(
      `react-native-obfuscating-transformer requires react-native >= 0.46`,
    )
  }
}

export interface ReactNativeObfuscatingTransformerDefaultResult {
  code: string
  map: string
}

export function maybeTransformMetroResult(
  upstreamResult: MetroTransformerResult,
  { code, map }: ReactNativeObfuscatingTransformerDefaultResult,
  reactNativeMinorVersion: number = getReactNativeMinorVersion(),
): MetroTransformerResult {
  if (reactNativeMinorVersion >= 52) {
    // convert code and map to ast
    const ast = babylon.parse(code, {
      sourceType: "module",
    })

    const mapConsumer = new SourceMapConsumer(map as any) // upstream types are wrong
      ; (traverse as any).cheap(ast, (node: Node) => {
        if (node.loc) {
          const originalStart = mapConsumer.originalPositionFor(node.loc.start)
          if (originalStart.line) {
            node.loc.start.line = originalStart.line
            node.loc.start.column = originalStart.column
          }
          const originalEnd = mapConsumer.originalPositionFor(node.loc.end)
          if (originalEnd.line) {
            node.loc.end.line = originalEnd.line
            node.loc.end.column = originalEnd.column
          }
        }
      })

    return { ast }
  } else if (Array.isArray(upstreamResult.map)) {
    return { code, map: convertStandardSourceMapToMetroRawSourceMap(map) }
  } else {
    return { code, map }
  }
}
