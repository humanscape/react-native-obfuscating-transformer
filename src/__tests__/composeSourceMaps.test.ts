import { getMetroTransformer } from "../getMetroTransformer"
import {
  convertMetroRawSourceMapToStandardSourceMap,
  MetroRawSourceMap,
  composeSourceMaps,
} from "../composeSourceMaps"
import * as fs from "fs"
import * as path from "path"
import { SourceMapConsumer } from "source-map/source-map"
import { getPositionOfSubstring } from "./getPositionOfSubstring"
import * as ts from "typescript"

const numberedLines = fs
  .readFileSync(require.resolve("./files/numberedLines.js"))
  .toString()

describe("convertMetroRawSourceMapToStandardSourceMap", () => {
  it("takes a raw source map and converts it to a non-raw source map", () => {
    const transformer = getMetroTransformer(47)

    const { map, code } = transformer.transform({
      filename: require.resolve("./files/numberedLines.js"),
      src: numberedLines,
      options: {
        retainLines: false,
      },
    })

    if (typeof code !== "string") {
      // use this rather than expect for typescript's sake
      throw new Error("code must be a string")
    }

    expect(Array.isArray(map)).toBe(true)
    expect(map).toMatchSnapshot()

    const standardMap = convertMetroRawSourceMapToStandardSourceMap(
      map as MetroRawSourceMap,
      path.join(__dirname, "numberedLines.js"),
      numberedLines,
    )

    const standardMapConsumer = new SourceMapConsumer(standardMap as any) // upstream types are wrong

    for (const substring of ["line1", "line2", "line3", "line5"])
      expect(
        standardMapConsumer.originalPositionFor(
          getPositionOfSubstring(code, substring)!,
        ),
      ).toMatchObject(getPositionOfSubstring(numberedLines, substring)!)

    expect(standardMap).toMatchSnapshot()
  })
})

describe("composeSourceMaps", () => {
  it("composes two source maps together", () => {
    const filename = require.resolve("./files/hello.ts")
    const hello = fs.readFileSync(filename).toString()

    const tsTranspileResult = ts.transpileModule(hello, {
      fileName: filename,
      compilerOptions: {
        sourceMap: true,
        target: ts.ScriptTarget.ES2015,
      },
    })

    expect(tsTranspileResult.outputText).toMatchSnapshot()

    const upstreamTransformResult = getMetroTransformer(47).transform({
      filename,
      src: tsTranspileResult.outputText,
      options: {
        retainLines: false,
      },
    })

    expect(upstreamTransformResult.code).toMatchSnapshot()

    const composedMap = new SourceMapConsumer(composeSourceMaps(
      tsTranspileResult.sourceMapText!,
      convertMetroRawSourceMapToStandardSourceMap(
        upstreamTransformResult.map as MetroRawSourceMap,
        filename,
        hello,
      ),
      filename,
      hello,
    ) as any) // upstream types are wrong

    expect(
      composedMap.originalPositionFor(
        getPositionOfSubstring(upstreamTransformResult.code!, "line6")!,
      ),
    ).toMatchObject({ line: 6 })

    expect(
      composedMap.originalPositionFor(
        getPositionOfSubstring(upstreamTransformResult.code!, "line8")!,
      ),
    ).toMatchObject({ line: 8 })
  })
})
