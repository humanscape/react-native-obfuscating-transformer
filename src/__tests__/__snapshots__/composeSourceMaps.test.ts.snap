// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composeSourceMaps composes two source maps together 1`] = `
"console.log(\\"line6\\");
export const line8 = true;
//# sourceMappingURL=hello.js.map"
`;

exports[`composeSourceMaps composes two source maps together 2`] = `
"Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
console.log(\\"line6\\");
var line8 = exports.line8 = true;"
`;

exports[`convertMetroRawSourceMapToStandardSourceMap takes a raw source map and converts it to a non-raw source map 1`] = `
Array [
  Array [
    4,
    0,
    1,
    7,
  ],
  Array [
    4,
    4,
    1,
    13,
    "line1",
  ],
  Array [
    4,
    28,
    1,
    21,
  ],
  Array [
    4,
    29,
    1,
    7,
  ],
  Array [
    5,
    0,
    2,
    7,
  ],
  Array [
    5,
    4,
    2,
    13,
    "line2",
  ],
  Array [
    5,
    28,
    2,
    21,
  ],
  Array [
    5,
    29,
    2,
    7,
  ],
  Array [
    6,
    0,
    3,
    7,
  ],
  Array [
    6,
    4,
    3,
    13,
    "line3",
  ],
  Array [
    6,
    28,
    3,
    21,
  ],
  Array [
    6,
    29,
    3,
    7,
  ],
  Array [
    8,
    0,
    5,
    7,
  ],
  Array [
    8,
    4,
    5,
    13,
    "line5",
  ],
  Array [
    8,
    28,
    5,
    21,
  ],
  Array [
    8,
    29,
    5,
    7,
  ],
]
`;

exports[`convertMetroRawSourceMapToStandardSourceMap takes a raw source map and converts it to a non-raw source map 2`] = `"{\\"version\\":3,\\"sources\\":[\\"/Users/dshe/code/react-native-obfuscating-transformer/src/__tests__/numberedLines.js\\"],\\"names\\":[\\"line1\\",\\"line2\\",\\"line3\\",\\"line5\\"],\\"mappings\\":\\";;;AAAO,IAAMA,wBAAQ,CAAd;AACA,IAAMC,wBAAQ,CAAd;AACA,IAAMC,wBAAQ,CAAd;;AAEA,IAAMC,wBAAQ,CAAd\\",\\"sourcesContent\\":[\\"export const line1 = 1\\\\nexport const line2 = 2\\\\nexport const line3 = 3\\\\n\\\\nexport const line5 = 5\\\\n\\"]}"`;
